### ISSUE TITLE
Can't get DelegateDecompiler to play with DateTime

### ISSUE THREAD

**Issue by Heras** [hazzik/DelegateDecompiler#85]
I absolutely love the concept, but I can't get it to work with comparing DateTimes, something that I need to do a lot.

I've tried

```
[Computed]
public bool IsOverdue
{
   return this.FinishedAt.HasValue && this.FinishedAt.Value < DateTime.Now.AddDays(-1);
}

```
but it throws

> LINQ to Entities does not recognize the method 'System.DateTime AddDays(Double)' method, and this method cannot be translated into a store expression.

I've tried using DbFunctions:

```
[Computed]
public bool IsOverdue
{
   return this.FinishedAt.HasValue && this.FinishedAt.Value < DbFunctions.AddDays(DateTime.Now, -1)
}
```

But it throws

> Unable to cast object of type 'System.Reflection.RuntimeConstructorInfo' to type 'System.Reflection.MethodInfo'.

Is there any way I can get a property like this to work, or am I out of luck?

**Comment by hazzik** [comment]
I think I know what the issue is, but I'll need to verify that. I'll take a look tomorrow.

**Comment by Heras** [comment]
Great! Thank you for your quick reply!

**Comment by hazzik** [comment]
@Heras what version do you use? Also, what .NET framework version? I can not reproduce on 0.19.0

**Comment by Heras** [comment]
Both latest, EntityFramework v6.1.3 & DelegateDecompiler v0.19.0

But your question triggered my memory: there may be some things wrong in our setup. I'm going to check that first!

**Comment by hazzik** [comment]
What about visual studio and .NET (not entity framework) version?

**Comment by Heras** [comment]
VS2015 Update 3 & .NET 4.6

**Comment by Heras** [comment]
We are still working with ObjectContext instead of DbContext. Could that be it?

**Comment by hazzik** [comment]
It is definetely about DbFunctions.AddDays. But dont know who generated "call <constructor>" in this case.

**Comment by Heras** [comment]
I'm afraid I don't fully understand your reply... Is it something that my code does, or something I can assist you with, or maybe a limit to DelegateDecompiler that it just cannot handle DateTime comparisons?

**Comment by Heras** [comment]
Hi hazzik, any way this problem can be resolved, or I can assist you? Otherwise I'm forced to move away from the DelegateDecompiler approach :(

**Comment by hazzik** [comment]
Can you please check the latest version? It might be fixed there.

**Comment by Heras** [comment]
I'm at v.0.19.0 already, but I did notice I have some nuget packages that have updates. I'll try again when I have every package up-to-date

**Comment by mwillebrands** [comment]
If you can't get it to work, can you create a demo project to demonstrate the issue?

**Comment by Heras** [comment]
[DecompileDateTime.zip](https://github.com/hazzik/DelegateDecompiler/files/508788/DecompileDateTime.zip)

The smallest project I could make that has the exact same problem. Just run and see it crash. So, at least I can be sure that it isn't caused by our "big" solution.

**Comment by Heras** [comment]
What might be causing it:

There's tricks to make relative DateTimes to work with LINQ to SQL:

- create a local var holding the new date
- use DbFunctions

I think in both cases, when compiled, the trick is "removed" from the expression, and decompiling will reintroduce the original problem that LINQ 2 SQL cannot translate DateTime.Now or DateTime.Today

**Comment by hazzik** [comment]
Hi @Heras 

> * create a local var holding the new date

This will not work as the variable will be inlined into the expression.

> * use DbFunctions

This works in the following way:
```
DbFunctions.AddYears(DateTime.Now, -10).Value
```
or
```
(DateTime) DbFunctions.AddYears(DateTime.Now, -10)
```
The `.Value`/cast is required because the DelegateDecompiler does not yet support optimizing folding of nullable value in this case.

**Comment by Heras** [comment]
Hi hazzik,

Fantastic!

AddYears returning a Nullable is such unexpected behaviour that I hadn't even noticed it.

Thanks for your quick replies and help. I will rewrite a lot of read-only properties to use the decompiler know, and have way cleaner code than we used to have.

Thanks again!

**Comment by hazzik** [comment]
Fixed in 9e114ff

**Comment by hazzik** [comment]
Pushed to NuGet
